core/server/proxy/src/main/java/alluxio/proxy/S3RestUtils.java

```diff
@@ -0,0 +1,283 @@
+/*
+ * The Alluxio Open Foundation licenses this work under the Apache License, version 2.0
+ * (the "License"). You may not use this work except in compliance with the License, which is
+ * available at www.apache.org/licenses/LICENSE-2.0
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.proxy;
+
+import alluxio.security.LoginUser;
+import alluxio.security.authentication.AuthenticatedClientUser;
+import alluxio.util.SecurityUtils;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+
+import javax.ws.rs.core.Response;
+import javax.xml.bind.annotation.XmlElement;
+import javax.xml.bind.annotation.XmlRootElement;
+
+/**
+ * Utilities for handling S3 REST calls.
+ */
+public final class S3RestUtils {
+  private static final Logger LOG = LoggerFactory.getLogger(S3RestUtils.class);
+
+  /**
+   * Calls the given {@link S3RestUtils.RestCallable} and handles any exceptions thrown.
+   *
+   * @param <T> the return type of the callable
+   * @param resource the resource (bucket or object) to be operated on
+   * @param callable the callable to call
+   * @return the response object
+   */
+  public static <T> Response call(String resource, S3RestUtils.RestCallable<T> callable) {
+    try {
+      if (SecurityUtils.isSecurityEnabled() && AuthenticatedClientUser.get() == null) {
+        AuthenticatedClientUser.set(LoginUser.get().getName());
```

Comment:
Setting `AuthenticatedClientUser` here does not seem right to me... `AuthenticatedClientUser` only makes sense and should only be used on the Alluxio Server JVM. Since proxy is still an Alluxio client to FSMaster, we should probably set the client-side `LoginUser` here. You can set `alluxio.security.login.username` temporality and reset it after the call is done.

Since this first prototype does not include security, I'm fine with marking it as a TODO here for now.